src/stim/util_top/reference_sample_tree.inl

```diff
@@ -0,0 +1,161 @@
+            auto &child = result.suffix_children.back();
+            child.repetitions = 1;
+            child.prefix_bits.insert(
+                child.prefix_bits.end(),
+                sim.measurement_record.storage.begin() + start_size,
+                sim.measurement_record.storage.begin() + end_size);
+        }
+        start_size = end_size;
+    };
+
+    for (size_t k = 0; k < reps; k++) {
+        for (const auto &inst : loop.operations) {
+            if (inst.gate_type == GateType::REPEAT) {
+                uint64_t repeats = inst.repeat_block_rep_count();
+                const auto& block = inst.repeat_block_body(loop);
+                flush_recorded_into_result();
+                result.suffix_children.push_back(do_loop_with_tortoise_hare_folding(block, repeats));
+                start_size = sim.measurement_record.storage.size();
+            } else {
+                sim.do_gate(inst);
+            }
+        }
+    }
+
+    flush_recorded_into_result();
+    return result;
+}
+
+static uint64_t max_feedback_lookback_in_loop(const Circuit &loop) {
+    uint64_t furthest_lookback = 0;
+    for (const auto &inst : loop.operations) {
+        if (inst.gate_type == GateType::REPEAT) {
+            furthest_lookback = std::max(furthest_lookback, max_feedback_lookback_in_loop(inst.repeat_block_body(loop)));
+        } else {
+            auto f = GATE_DATA[inst.gate_type].flags;
+            if ((f & GateFlags::GATE_CAN_TARGET_BITS) && (f & GateFlags::GATE_TARGETS_PAIRS)) {
+                // Feedback-capable operation. Check for any measurement record targets.
+                for (auto t : inst.targets) {
+                    if (t.is_measurement_record_target()) {
+                        furthest_lookback = std::max(furthest_lookback, (uint64_t)-t.rec_offset());
+                    }
+                }
+            }
+        }
+    }
+    return furthest_lookback;
+}
+
+template <size_t W>
+ReferenceSampleTree CompressedReferenceSampleHelper<W>::do_loop_with_tortoise_hare_folding(const Circuit &loop, uint64_t reps) {
+    if (reps < 10) {
+        return do_loop_with_no_folding(loop, reps);
+    }
+
+    ReferenceSampleTree result;
+    result.repetitions = 1;
+
+    CompressedReferenceSampleHelper<W> tortoise(sim);
+    CompressedReferenceSampleHelper<W> hare(std::move(sim));
+    uint64_t max_feedback_lookback = max_feedback_lookback_in_loop(loop);
+    uint64_t tortoise_steps = 0;
+    uint64_t hare_steps = 0;
+    while (hare_steps < reps) {
+        hare_steps++;
+        result.suffix_children.push_back(hare.do_loop_with_no_folding(loop, 1));
+        assert(result.suffix_children.size() == hare_steps);
+
+        if (tortoise.in_same_recent_state_as(hare, max_feedback_lookback, hare_steps < 10)) {
+            break;
+        }
+
+        // Tortoise advances half as quickly.
+        if (hare_steps & 1) {
+            tortoise_steps++;
+            tortoise.do_loop_with_no_folding(loop, 1);
```

Comment:
The tortoise costs at most 50% of what the hare costs, so although this is inefficient it's boundedly bad. Will revisit if we need more performance later. I'm more worried about the performance of the `canonical_stabilizers` calls than I am about this.

Reply:
Something counterintuitive here is that it looks like the overhead is exponential under nesting, like you gain 1.5x in cost per level of nesting, but because iterations must also be increasing exponentially with nesting (by at least 10x since otherwise this code takes the escape route avoiding tortoise-and-hare) this 1.5x is not so fatal. It's sort of like... instead of growing like 10^n it will grow like 1.5^n.

If we do ever care about circuits with extreme nesting, it would be worth adding a caching mechanism to avoid constantly resolving each loop.

Reply:
> The tortoise costs at most 50% of what the hare costs

Yes, but there's also `TableauSimulator` copying and other overhead that is just unnecessary that is happening recursively.

> I'm more worried about the performance of the `canonical_stabilizers` calls than I am about this.

Again, all of that is called recursively.

> it would be worth adding a caching mechanism to avoid constantly resolving each loop.

Caching would definitely be useful, especially when loop iterations get long and more nested. However, I think the simple thing is to just not do the extra recursive work that is thrown away.

Reply:
In general, a loop iteration will tend to touch the entire tableau's state (e.g. a layer of Hadamard gates will do this). So I expect the copies are less expensive than simulating an iteration.

Anyways... I agree it could be better, but it's IMO pretty complex to make better and I am going to wait until forced by a benchmark to do so.